### ISSUE TITLE
bad performance for setting resnet20 as a student and resnet56 as a teacher

### ISSUE THREAD

**Issue by Slawlight** [clovaai/overhaul-distillation#13]
When I set resnet20 as a student and resnet56 as a teacher on cifar-100, it got a very poor result. Setting: lr=0.05, batch-size=128, weight-decay=5e-4, alpha=1e-3.

**Comment by bhheo** [comment]
Hi

Could you please share me more details about your result ?
such as teacher accuracy, student accuracy w/o distill, with distill.
Because I didn't test it on that case, I need more information.
If performance is lower than baseline, I recommend you to use smaller alpha such as 1e-4.

**Comment by Slawlight** [comment]
baseline:  resnet-110 test_acc 72.06    resnet-20 test_acc 68.95.   
Total training epochs: 150. 
The initial learning rate is 0.05 and decay at 80 and 120 epoch by multiply 0.1 , the optimizer is SGD with momentum 0.9.
the student test accuracy with distill: <65. I didn't finish the distillation process, because it is bad at the beginning.

**Comment by Slawlight** [comment]
Oh , I make a mistake about the discription. The teacher is resnet56.  Just replace 110 with 56.

**Comment by Slawlight** [comment]
like this:
<img width="622" alt="屏幕快照 2019-11-23 下午4 58 30" src="https://user-images.githubusercontent.com/18410535/69476243-8ca44f00-0e12-11ea-83f5-0f7255b33a28.png">

**Comment by bhheo** [comment]
OK
Something is wrong in that setting.
Because this setting is smaller than my smallest setting in paper, I think my algorithm can be poor at this small scale.
But, I'm not sure about that.

Could you share your code just temporarily using github or online drive ?
To be sure, I want to check the code.

**Comment by Slawlight** [comment]
The code : https://github.com/Slawlight/overhaul.git

**Comment by bhheo** [comment]
Thanks for your sharing.
I have checked your codes and finds two problems.
you can find changes [here](https://github.com/bhheo/overhaul/commit/0bf9fa5ab8996963bda96fb5ba64e9d8bf877372)

First, one missing ReLU in forward function of ResNet.
You have to add ```x = F.relu(x)``` at line 81 of cifar_resnet.py
Second, you have to train parameters of connector layers in distiller.
The optimizer have to be changed as 
```Python
optimizer = optim.SGD([{'params': s_net.parameters()}, {'params': d_net.Connectors.parameters()}], 
args.lr, momentum=args.momentum, weight_decay=args.weight_decay)
```

As a results, accuracy of my distillation was changed as follows
1. Original : 67.2%
2. with ReLU : 68.78%
3. with optim on connectors : 70.06% 

Because of the small network size, it's not very effective but shows similar performance to KD.
I also recommend you to use ```nesterov=True``` option on ```optim.SGD``` to make training more fast and stable.

**Comment by Slawlight** [comment]
It works. Thank you very much.
